### PR TITLE
Fix WooCommerce sync performance on big lists [MAILPOET-1983]

### PR DESCRIPTION
**Pull request by wxa**
Two issues which affected WC sync performance were fixed: 

* Slow queries in the `removeOrphanedSubscribers()` method:
  * The WC customer data we need is stored in `postmeta` and `usermeta` key-value tables, the values aren't indexed, so any big join on these tables is bound to be slow. We now first pull the data into temporary tables which have indexes, thus making joins cheaper;
  * `LEFT JOIN` seems not to use indexes when joined columns (`email` in this case) have different charsets, so for fields in temporary tables we need to set a charset and collation that match our subscribers table.
* The scheduled task was buggy: 
  * There was a needless `CronHelper` timeout check after sync completion so if sync took too long the task was never marked as `completed`;
  * Due to the above and general slowness of the sync the task could run in multiple instances at once, further slowing down the host. I've implemented a flag to not run more than one instance.

Please use a sufficiently large customer list (~30K) for testing. You may ask @qfrery for a sample dump.

### CODE REVIEW

File: lib/Cron/Workers/WooCommerceSync.php [hunk 1]
```diff
@@ -27,13 +29,22 @@ function checkProcessingRequirements() {
...
+    if (is_null($task->status)) {
+      $this->taskAlreadyInProgress = true;
+      return false;
+    }
+
+    return parent::prepareTask($task);
+  }
+
   function processTaskStrategy(ScheduledTask $task) {
+    if ($this->taskAlreadyInProgress) {
```
Comment (costasovo) [inline lib/Cron/Workers/WooCommerceSync.php]: What will happen if the task fails (e.g. some mysql timeout or just some server memory issue)? I'm afraid that this task could be stuck in running state. Do we have some mechanism to mark task as timeouted and reschedule it?
I was not able to find something like that in SimpleWorker. Maybe we could add some execution timeout e.g. one day and in case the task is running for a day (is null and not updated for a day) then run it again and update tasks updated_at

**Comment by costasovo** [comment]
I tried to simulate a big WC site. I've generated 30K customers and 30K guest orders then I ran a cron job to sync all to WC Segment. My setup is not very fast so I suppose that on some server this would run faster.

**TLDR:** I would say that it is ok. It took some time but site was responsive during the process. Maybe we could optionally improve deleting orphans.

I've uploaded test site dump and complete query log to [JIRA](https://mailpoet.atlassian.net/browse/MAILPOET-1983)  in case you are interested.

### SYNC IN WC LIST:

Cron ran approx a minute and site was responsive.
Here are slowest queries I logged during the run (time is in seconds):

```sql
      UPDATE wp_mailpoet_subscribers mps
        JOIN wp_postmeta wppm ON mps.email = wppm.meta_value AND wppm.meta_key = "_billing_email"
        JOIN wp_postmeta wppm2 ON wppm2.post_id = wppm.post_id AND wppm2.meta_key = "_billing_first_name"
        JOIN (SELECT MAX(post_id) AS max_id FROM wp_postmeta WHERE meta_key = "_billing_email" GROUP BY meta_value) AS tmaxid ON tmaxid.max_id = wppm.post_id
      SET mps.first_name = wppm2.meta_value
        WHERE mps.first_name = ""
        AND mps.is_woocommerce_user = 1
        AND wppm2.meta_value IS NOT NULL
    , 19.362863063812 

	  UPDATE wp_mailpoet_subscribers mps
        JOIN wp_postmeta wppm ON mps.email = wppm.meta_value AND wppm.meta_key = "_billing_email"
        JOIN wp_postmeta wppm2 ON wppm2.post_id = wppm.post_id AND wppm2.meta_key = "_billing_last_name"
        JOIN (SELECT MAX(post_id) AS max_id FROM wp_postmeta WHERE meta_key = "_billing_email" GROUP BY meta_value) AS tmaxid ON tmaxid.max_id = wppm.post_id
      SET mps.last_name = wppm2.meta_value
        WHERE mps.last_name = ""
        AND mps.is_woocommerce_user = 1
        AND wppm2.meta_value IS NOT NULL
    , 5.9080939292908 
```  

### SYNC OUT WC LIST:

After that I tried deleted all billing emails and removed `customer` capability from all users and ran sync again.
I this (extreme) case it took cron to 15 minutes to finish.

The queries performed same as in first usecase. Most time was spent updating/deleting subscribers from segment.

It was caused by usage of:

```php
	->findResultSet()
	->set('is_woocommerce_user', 0)
	->save();

	// and

	->findResultSet()
    ->set('is_woocommerce_user', 0)
    ->delete();
```
This fetches data and then perform update (or update + delete) one by one. I think this could be improved by fetching ids and then calling several batch updates. But this test was an extreme usecase and it finished correctly just took a lot of time so maybe there is no need to act on this.

File: lib/Cron/Workers/WooCommerceSync.php (hunk 1 above)
Comment (wxa) [reply lib/Cron/Workers/WooCommerceSync.php]: We don't have any mechanism for dealing with faulty tasks. I think a global mechanism is out of the scope of this issue (we need to consider its impact on all other tasks), but I've added a timeout check inside the WC sync task.

P.S.: I also found that the previous solution for ensuring a single instance didn't work because the flag was a class property, but `prepareTask` and `processTask` were called in different requests so the flag was never passed from one to another. Now I've added a `meta` field to scheduled tasks (like in sending queues) for persistent storage of this kind of data.

**Comment by wxa** [comment]
Thank you for thorough testing. Although the sync out case is an edge case, we could improve it a bit. For the updates it should be relatively easy, but deletions are harder because the `delete()` model method performs a few queries to remove associated data. If we rewrite them in batch we may have to maintain two versions of the same operation.

File: lib/Cron/Workers/WooCommerceSync.php [hunk 2]
```diff
@@ -29,22 +29,23 @@ function checkProcessingRequirements() {
...
-      return false;
-    }
-
-    return parent::prepareTask($task);
-  }
-
   function processTaskStrategy(ScheduledTask $task) {
-    if ($this->taskAlreadyInProgress) {
+    $meta = $task->getMeta();
+    if (!empty($meta['in_progress'])) {
```
Comment (costasovo) [inline lib/Cron/Workers/WooCommerceSync.php]: Do we need extra meta field for this information? I think that the same information is stored in `status` (`$task->status === null`), isn't it?

Comment (wxa) [reply lib/Cron/Workers/WooCommerceSync.php]: Not quite. The `processTask` method is called on tasks with `status = null` on every cron run. We can't distinguish between a first and non-first run by the status itself, so I've added a flag for this purpose.

Comment (costasovo) [reply lib/Cron/Workers/WooCommerceSync.php]: Ahh I see. I always thought that prepareTask and porcessTask are called within one run :(

Comment (wxa) [reply lib/Cron/Workers/WooCommerceSync.php]: Yeah, I forgot this too when I did the first version of the fix. This was done because some tasks require expensive preparation (e.g. populating subscriber queues).

**Comment by wxa** [comment]
@michelleshull I have made two more changes:
* Fixed the "WC import page displayed" flag being reset when saving WC settings;
* Removed the feature flag. Now we have all past disabled WC features active, please check that they work well in conjunction with each other. After you finish, please assign the issue to Quentin for a final test.
